drgn_tools/crash_net.py

```diff
@@ -0,0 +1,372 @@
+            ifa = ifa.ifa_next
+    return ips
+
+
+def netdev_ipv6s(dev: Object) -> List[ipaddress.IPv6Address]:
+    """
+    Return a list of IPv6 addresses associated with a net_device.
+
+    :param dev: ``struct net_device *``
+    :returns: list of IPv6 addresses
+    """
+    ips = []
+    prog = dev.prog_
+    if dev.ip6_ptr:
+        for addr in list_for_each_entry(
+            "struct inet6_ifaddr",
+            dev.ip6_ptr.addr_list.address_of_(),
+            "if_list",
+        ):
+            addr_bytes = prog.read(addr.addr.in6_u.u6_addr8.address_of_(), 16)
+            ips.append(ipaddress.IPv6Address(addr_bytes))
+    return ips
+
+
+def for_each_netdev(ns: Object) -> Iterator[Tuple[str, Object]]:
+    """
+    Iterate over each ``struct net_device *`` of a net namespace
+
+    :param ns: ``struct net *``
+    :returns: an iterator of ``struct net_device *``
+    """
+    # Since commit ff92741270bf ("net: introduce name_node struct
+    # to be used in hashlist"), net.dev_name_head is a list of name nodes. Prior
+    # to that, it contains the net_devices directly.
+    # The below approach is similar to what's already in drgn's
+    # netdev_get_by_name().
+    try:
+        entry_type = ns.prog_.type("struct netdev_name_node")
+        member = "hlist"
+        entry_is_name_node = True
+    except LookupError:
+        entry_type = ns.prog_.type("struct net_device")
+        member = "name_hlist"
+        entry_is_name_node = False
+
+    for i in range(_NETDEV_HASHENTRIES):
+        head = ns.dev_name_head[i]
+        for dev_node in hlist_for_each_entry(entry_type, head, member):
+            dev = dev_node.dev if entry_is_name_node else dev_node
+            yield (dev_node.name.string_().decode("utf-8"), dev)
+
+
+def net_crash(prog: Program) -> None:
+    """
+    Display the network devices, their names and addresses.
+    :param prog: drgn program
+    :returns: None.
+    """
+    rows = [["NET_DEVICE", "NAME", "IP ADDRESS(ES)"]]
+
+    net_namespaces = network.for_each_net(prog)
+    for net_namespace in net_namespaces:
+        for name, dev in for_each_netdev(net_namespace):
+            net_device_ptr = hex(dev.dev.platform_data)
+            ip_list = [str(a) for a in netdev_ipv4s(dev) + netdev_ipv6s(dev)]
+            ips = ", ".join(ip_list)
+            rows.append(
+                [
+                    net_device_ptr,
+                    name,
+                    ips,
+                ]
+            )
+    print_table(rows)
+
+
+def crash_a(prog: Program) -> None:
+    """
+    Displays the arp cache.
+    :param prog: drgn program
+    :returns: None.
+    """
+
+    rows = [
+        ["NEIGHBOUR", "IP ADDRESS", "HW TYPE", "HW ADDRESS", "DEVICE", "STATE"]
+    ]
+
+    nht = prog["arp_tbl"].nht
+
+    hash_shift = nht.hash_shift.value_()
+    nhash_buckets = 1 << hash_shift
+    hash_buckets = nht.hash_buckets
+
+    for i in range(nhash_buckets):
+        neigh = hash_buckets[i]
+        while neigh:
```

Comment:
Re-iterating another comment from before:
It would be nice if you could factor out this part into a `for_each_neighbor(prog)` function!